sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -68,30 +68,25 @@ Providing a generic group abstraction encapsulating the shared functionality of
 
 === Attributes
 
-The +[[intel::reqd_sub_group_size(n)]]+ attribute indicates that the kernel must be compiled and executed with a sub-group of size _n_.  The value of _n_ must be a compile-time integral constant expression.  The value of _n_ must be set to a sub-group size that is both supported by the device and compatible with all language features used by the kernel, or device compilation will fail.  The set of valid sub-group sizes can be queried as described below.
+The +[[intel::sub_group_size(S)]]+ attribute indicates that the kernel must be compiled and executed with a specific sub-group size.  The value of _S_ must be a compile-time integral constant expression.  The kernel should only be submitted to a device that supports that sub-group size (as reported by +info::device::sub_group_sizes+).  If the kernel is submitted to a device that does not support the requested sub-group size, or a device on which the requested sub-group size is incompatible with any language features used by the kernel, the implementation must throw a synchronous exception with the `errc::feature_not_supported` error code from the kernel invocation command.
 
-In addition to device functions, the required sub-group size attribute may also be specified in the definition of a named functor object, as in the example below:
+The +[[intel::named_sub_group_size(NAME)]]+ attribute indicates that the kernel must be compiled and executed with a named sub-group size.  _NAME_ must be one of the following special tokens: +auto+, +primary+.  If _NAME_ is +auto+, the implementation is free to select any of the valid sub-group sizes associated with the device to which the kernel is submitted; the manner in which the sub-group size is selected is implementation-defined.  If _NAME_ is +primary+, the implementation will select the device's primary sub-group size (as reported by the +info::device::primary_sub_group_size+ query) for all kernels with this attribute.
 
-[source, c++]
-----
-class Functor
-{
-  void operator()(item<1> item) [[intel::reqd_sub_group_size(16)]]
-  {
-    /* kernel code */
-  }
-}
-----
+If no sub-group size attribute appears on a kernel, the default behavior is as-if +[[intel::named_sub_group_size(auto)]]+ was specified.  This behavior may be overridden by an implementation (e.g. via compiler flags).
+
+Sub-group size attributes may also be applied to `SYCL_EXTERNAL` functions.  If a kernel calls a `SYCL_EXTERNAL` function, or a `SYCL_EXTERNAL` function calls another `SYCL_EXTERNAL` function, the attributes applied to the caller and callee must match exactly.  If the attributes do not match, the compiler should produce an error.  Note that sub-group size attributes are not propagated from a device function to callers of the function, and must be specified explicitly.
```

Comment:
I found this paragraph unclear for several reasons:

* It's not clear if the sub-group size attribute is **required** in this scenario or if it is merely **allowed**.  (I think we want it to be required, correct?)

* Whenever `SYCL_EXTERNAL` is used, there are two relevant translation units: the TU the makes the call and the TU that defines the function.  We need to make it clear whether the attribute is required in the calling TU, the defining TU, or both.

* The statement about requiring the compiler to produce an error makes it sound like the compiler must do inter-TU analysis to see if the function defined via `SYCL_EXTERNAL` is defined with the right attribute.  (Or at least the statement could be interpreted that way.)  I think that is not our intent.

* The statement about the sub-group size not be propagated from a device function to its callers could be interpreted to mean that this propagation doesn't happen even within a single TU.  I think the intent, though, is that this propagation does not happen across TUs.

Does the following paragraph capture what we want to say?

> There are special requirements whenever a device function defined in one translation unit makes a call to a device function that is defined in a second translation unit.  In such a case, the second device function is always declared using `SYCL_EXTERNAL`.  If the kernel containing these device function is defined using a sub-group size attribute, the functions declared using `SYCL_EXTERNAL` must also be decorated with that same attribute.  This decoration must exist in both the translation unit making the call and also in the translation unit that defines the function.  If the sub-group attribute is missing in the translation unit that makes the call (or if the sub-group size of the called function does not match the sub-group size of the calling function), the program is ill formed and the compiler must raise a diagnostic.

Reply:
Yes, this is much better, thanks.  I added a note that `auto` doesn't make sense on a `SYCL_EXTERNAL` function.  See bd68110bd403386c54a6c3e749403a741578e687.